### PR TITLE
[SPARK-5479] [yarn] Handle --py-files correctly in YARN.

### PR DESCRIPTION
**Pull request by vanzin**
The bug description is a little misleading: the actual issue is that
.py files are not handled correctly when distributed by YARN. They're
added to "spark.submit.pyFiles", which, when processed by context.py,
explicitly whitelists certain extensions (see PACKAGE_EXTENSIONS),
and that does not include .py files.

On top of that, archives were not handled at all! They made it to the
driver's python path, but never made it to executors, since the mechanism
used to propagate their location (spark.submit.pyFiles) only works on
the driver side.

So, instead, ignore "spark.submit.pyFiles" and just build PYTHONPATH
correctly for both driver and executors. Individual .py files are
placed in a subdirectory of the container's local dir in the cluster,
which is then added to the python path. Archives are added directly.

The change, as a side effect, ends up solving the symptom described
in the bug. The issue was not that the files were not being distributed,
but that they were never made visible to the python application
running under Spark.

Also included is a proper unit test for running python on YARN, which
broke in several different ways with the previous code.

A short walk around of the changes:
- SparkSubmit does not try to be smart about how YARN handles python
  files anymore. It just passes down the configs to the YARN client
  code.
- The YARN client distributes python files and archives differently,
  placing the files in a subdirectory.
- The YARN client now sets PYTHONPATH for the processes it launches;
  to properly handle different locations, it uses YARN's support for
  embedding env variables, so to avoid YARN expanding those at the
  wrong time, SparkConf is now propagated to the AM using a conf file
  instead of command line options.
- Because the Client initialization code is a maze of implicit
  dependencies, some code needed to be moved around to make sure
  all needed state was available when the code ran.
- The pyspark tests in YarnClusterSuite now actually distribute and try
  to use both a python file and an archive containing a different python
  module. Also added a yarn-client tests for completeness.
- I cleaned up some of the code around distributing files to YARN, to
  avoid adding more copied & pasted code to handle the new files being
  distributed.

### CODE REVIEW

**Comment by vanzin** [comment]
Tests run:
- unit tests against hadoop 2.5.0
- real yarn cluster, client and cluster modes (CDH 5.4). In cluster mode, I used a .py file stored in HDFS.
- standalone client mode
- local mode

**Comment by vanzin** [comment]
Jenkins, retest this please.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -372,11 +381,29 @@ private[spark] class Client(
       sparkConf.set(CONF_SPARK_YARN_SECONDARY_JARS, cachedSecondaryJarLinks.mkString(","))
     }
 
+    if (args.primaryPyFile != null) {
+      distribute(args.primaryPyFile)
```
Comment (lianhuiwang) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: i think primaryPyFile is only used in applicationMaster, so it is distributed to appMaster.

**Comment by srowen** [comment]
I'm not an expert on this code by any means. The problem description and solution sound reasonable and match up with the change. I skimmed the change and didn't see anything I'd flag. Tests pass. I'd love to get more Python folks to look at it, like @davies (and I see @lianhuiwang is already looking too), but this appears to be a good cleanup and problem solving change for 1.5 / `master`.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Don't you need the file around when you run code in the executors? How is the code shipped to executors otherwise?

I can try it out, but it feels weird to have used code only in the AM.

Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Hmm, strangely enough it seems things work without distributing the main file. Wonder if that's safe in all cases or just a side-effect of the really simple test code, though.

**Comment by vanzin** [comment]
Ping @tgravescs @andrewor14

**Comment by tgravescs** [comment]
@vanzin  sorry I'm going to be on vacation for the next week

**Comment by vanzin** [comment]
jenkins, retest this please.

**Comment by srowen** [comment]
@tgravescs shall we give you a few more days to take a look? Tests pass and I tend to trust Mr Vanzin here. @davies do you have any input from a Python perspective?

**Comment by andrewor14** [comment]
Hey I'll take a look at this later today, sorry for slipping.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 2]
```diff
@@ -315,62 +315,59 @@ object SparkSubmit {
...
-        if (!sys.env.contains("SPARK_HOME")) {
-          printErrorAndExit("SPARK_HOME does not exist for python application in yarn mode.")
-        }
-        val pythonPath = new ArrayBuffer[String]
-        for (sparkHome <- sys.env.get("SPARK_HOME")) {
-          val pyLibPath = Seq(sparkHome, "python", "lib").mkString(File.separator)
-          val pyArchivesFile = new File(pyLibPath, "pyspark.zip")
-          if (!pyArchivesFile.exists()) {
-            printErrorAndExit("pyspark.zip does not exist for python application in yarn mode.")
+    if (args.isPython) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: I actually find the prior organization of this code a little easier to follow. It read
```
if (args.isPython && clusterManager == YARN) { /* do something */ }
if (args.isPython && deployMode == CLIENT) { /* do something else */ }
```
So if I only cared about python but not YARN I can just skip the first block and read the second block. Now it's one giant python block. Not a big deal, but if merging these isn't strictly necessary I would prefer the old structure.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: (also it'll be easier to back port stuff)

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 3]
```diff
@@ -315,62 +315,59 @@ object SparkSubmit {
...
-            printErrorAndExit("py4j-0.8.2.1-src.zip does not exist for python application " +
-              "in yarn mode.")
+          val pythonPath = new ArrayBuffer[String]
+          for (sparkHome <- sys.env.get("SPARK_HOME")) {
+            val pyLibPath = Seq(sparkHome, "python", "lib").mkString(File.separator)
+            val pyArchivesFile = new File(pyLibPath, "pyspark.zip")
+            if (!pyArchivesFile.exists()) {
+              printErrorAndExit("pyspark.zip does not exist for python application in yarn mode.")
+            }
+            val py4jFile = new File(pyLibPath, "py4j-0.8.2.1-src.zip")
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: not your code, but at some point we should factor this out instead of hardcoding the version here

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 4]
```diff
@@ -315,62 +315,59 @@ object SparkSubmit {
...
         } else {
-          localURI.getPath
+          // If a python file is provided, add it to the child arguments and list of files to
+          // deploy.
+          // Usage: PythonAppRunner <main python file> <extra python files> [app arguments]
+          args.mainClass = "org.apache.spark.deploy.PythonRunner"
+          args.childArgs = ArrayBuffer(args.primaryResource, args.pyFiles) ++ args.childArgs
+          if (clusterManager != YARN) {
+            args.files = mergeFileLists(args.files, args.primaryResource)
+          }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: can you add a comment to explain why we don't want to do this in YARN?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 5]
```diff
@@ -387,13 +384,6 @@ object SparkSubmit {
...
-      // In yarn-cluster mode for a python app, add primary resource and pyFiles to files
-      // that can be distributed with the job
-      if (args.isPython) {
-        args.files = mergeFileLists(args.files, args.primaryResource)
-        args.files = mergeFileLists(args.files, args.pyFiles)
-      }
-
       // In yarn-cluster mode for a R app, add primary resource to files
       // that can be distributed with the job
       if (args.isR) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: now that you removed the python stuff you can simplify this to `if (isYarnCluster && args.isR) {`

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 4 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: just so I didn't miss anything, the only functional changes in this block are that you added YARN checks here and in L364, right?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 6]
```diff
@@ -487,9 +495,9 @@ private[spark] class ApplicationMaster(
         new MutableURLClassLoader(urls, Utils.getContextOrSparkClassLoader)
       }
 
+    var userArgs = args.userArgs
     if (args.primaryPyFile != null && args.primaryPyFile.endsWith(".py")) {
-      System.setProperty("spark.submit.pyFiles",
-        PythonRunner.formatPaths(args.pyFiles).mkString(","))
+      userArgs = Seq(args.primaryPyFile, "") ++ userArgs
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: why the empty string?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 7]
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
...
+        resType: LocalResourceType = LocalResourceType.FILE,
+        destName: Option[String] = None,
+        targetDir: Option[String] = None,
+        appMasterOnly: Boolean = false): (Boolean, String) = {
+      val localURI = new URI(path.trim())
+      if (localURI.getScheme != LOCAL_SCHEME) {
+        if (addDistributedUri(localURI)) {
+          val localPath = getQualifiedLocalPath(localURI, hadoopConf)
+          val linkname = targetDir.map(_ + "/").getOrElse("") +
+            destName.orElse(Option(localURI.getFragment())).getOrElse(localPath.getName())
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: can you explain why we use the fragment here? can you add a comment?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 8]
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
...
     /**
+     * Distribute a file to the cluster.
+     *
+     * @param path URI of the file to distribute.
+     * @param resType Type of resource being distributed.
+     * @param destName Name of the file in the distributed cache.
+     * @param targetDir Subdirectory where to place the file.
+     * @param appMasterOnly Whether to distribute only to the AM.
+     * @return A 2-tuple. First item is whether the file is a "local:" URI. Second item is the
+     *         localized path for non-local paths, or the input `path` for local paths.
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: The second item can be null. Can you document when this is the case to avoid NPEs in the future if this isn't used carefully?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 9]
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
...
+
+    // If we passed in a keytab, make sure we copy the keytab to the staging directory on
+    // HDFS, and setup the relevant environment vars, so the AM can login again.
+    if (loginFromKeytab) {
+      logInfo("To enable the AM to login from keytab, credentials are being copied over to the AM" +
+        " via the YARN Secure Distributed Cache.")
+      val (_, localizedPath) = distribute(args.keytab,
+        destName = Some(sparkConf.get("spark.yarn.keytab")),
+        appMasterOnly = true)
+      require(localizedPath != null)
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: do we require here because the key tab should only be distributed once? If so can you add a comment about that here can maybe a fail message for the require?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 10]
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
...
+     * @param appMasterOnly Whether to distribute only to the AM.
+     * @return A 2-tuple. First item is whether the file is a "local:" URI. Second item is the
+     *         localized path for non-local paths, or the input `path` for local paths.
+     */
+    def distribute(
+        path: String,
+        resType: LocalResourceType = LocalResourceType.FILE,
+        destName: Option[String] = None,
+        targetDir: Option[String] = None,
+        appMasterOnly: Boolean = false): (Boolean, String) = {
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: thanks for cleaning this up with a helpful java doc!

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 11]
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
...
+     *
+     * @param path URI of the file to distribute.
+     * @param resType Type of resource being distributed.
+     * @param destName Name of the file in the distributed cache.
+     * @param targetDir Subdirectory where to place the file.
+     * @param appMasterOnly Whether to distribute only to the AM.
+     * @return A 2-tuple. First item is whether the file is a "local:" URI. Second item is the
+     *         localized path for non-local paths, or the input `path` for local paths.
+     */
+    def distribute(
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: can be private I think

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 12]
```diff
@@ -307,33 +342,18 @@ private[spark] class Client(
...
-          if (addDistributedUri(localURI)) {
-            val src = getQualifiedLocalPath(localURI, hadoopConf)
-            val destPath = copyFileToRemote(dst, src, replication)
-            val destFs = FileSystem.get(destPath.toUri(), hadoopConf)
-            distCacheMgr.addResource(destFs, hadoopConf, destPath,
-              localResources, LocalResourceType.FILE, destName, statCache)
-          }
-        } else if (confKey != null) {
+    ).foreach { case (destName, path, confKey) =>
+      if (path != null && !path.isEmpty()) {
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: this is very slightly different from before; before we trimmed the path before checking whether it's empty. Maybe it's not necessary but just to make sure we're aware of this

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 13]
```diff
@@ -307,33 +342,18 @@ private[spark] class Client(
...
+        val (isLocal, localizedPath) = distribute(path, destName = Some(destName))
+        if (isLocal && confKey != null) {
           // If the resource is intended for local use only, handle this downstream
           // by setting the appropriate property
-          sparkConf.set(confKey, localPath)
+          require(localizedPath != null)
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: the comment above does not describe this require, so we should move it down 1 line (also a fail message here would be good)

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 14]
```diff
@@ -349,21 +369,10 @@ private[spark] class Client(
...
-              distCacheMgr.addResource(
-                fs, hadoopConf, destPath, localResources, resType, linkname, statCache)
-              if (addToClasspath) {
-                cachedSecondaryJarLinks += linkname
-              }
-            }
-          } else if (addToClasspath) {
-            // Resource is intended for local use only and should be added to the class path
-            cachedSecondaryJarLinks += file.trim()
+          val (isLocal, localizedPath) = distribute(file, resType = resType)
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: isLocal is not used

Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: actually, shouldn't it be used?
```
if (isLocal && addToClassPath) { ... }
```

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 15]
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
...
+     *         localized path for non-local paths, or the input `path` for local paths.
+     */
+    def distribute(
+        path: String,
+        resType: LocalResourceType = LocalResourceType.FILE,
+        destName: Option[String] = None,
+        targetDir: Option[String] = None,
+        appMasterOnly: Boolean = false): (Boolean, String) = {
+      val localURI = new URI(path.trim())
+      if (localURI.getScheme != LOCAL_SCHEME) {
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: not your code, but can you add a comment here to explain the differences between how we handle resources with local schemes and ones without

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 16]
```diff
@@ -471,9 +505,6 @@ private[spark] class Client(
       val renewalInterval = getTokenRenewalInterval(stagingDirPath)
       sparkConf.set("spark.yarn.token.renewal.interval", renewalInterval.toString)
     }
-    // Set the environment variables to be passed on to the executors.
-    distCacheMgr.setDistFilesEnv(env)
-    distCacheMgr.setDistArchivesEnv(env)
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I noticed that we order some of these code blocks around in this patch. Is there a particular reason for it? Have you verified that it's safe to do so?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 17]
```diff
@@ -899,8 +938,21 @@ object Client extends Logging {
   // Distribution-defined classpath to add to processes
   val ENV_DIST_CLASSPATH = "SPARK_DIST_CLASSPATH"
 
-  // Subdirectory where the user's hadoop config files will be placed.
-  val LOCALIZED_HADOOP_CONF_DIR = "__hadoop_conf__"
+  // Subdirectory where the user's Spark and Hadoop config files will be placed.
+  val LOCALIZED_CONF_DIR = "__spark_conf__"
+
+  // Name fo the file in the conf archive containing Spark configuration.
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Name of

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 18]
```diff
@@ -899,8 +938,21 @@ object Client extends Logging {
...
+  val SPARK_CONF_FILE = "__spark_conf__.properties"
+
+  // Subdirectory where the user's python files (not archives) will be placed.
+  val LOCALIZED_PYTHON_DIR = "__pyfiles__"
+
+  // Key in SparkConf where to find the executors' PYTHONPATH. This cannot be set using
+  // `SparkConf.setExecutorEnv`, because that would cause it to propagate to the python
+  // code via the configuration, and then override the process's own environment when
+  // launching workers. Since it contains variables that are expanded by YARN, that cannot
+  // happen.
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: this is a little long-winded. Did you mean to say that there may be variables in this python path that are substituted on the executor side?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 19]
```diff
@@ -548,8 +596,13 @@ private[spark] class Client(
     logInfo("Setting up container launch context for our AM")
     val appId = newAppResponse.getApplicationId
     val appStagingDir = getAppStagingDir(appId)
-    val localResources = prepareLocalResources(appStagingDir)
     val launchEnv = setupLaunchEnv(appStagingDir)
+    val localResources = prepareLocalResources(appStagingDir)
+
+    // Set the environment variables to be passed on to the executors.
+    distCacheMgr.setDistFilesEnv(launchEnv)
+    distCacheMgr.setDistArchivesEnv(launchEnv)
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I noticed that we order some of these code blocks around in this patch. Is there a particular reason for it? Have you verified that it's safe to do so?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 20]
```diff
@@ -490,15 +521,32 @@ private[spark] class Client(
...
-    // if spark.submit.pyArchives is in sparkConf, append pyArchives to PYTHONPATH
-    // that can be passed on to the ApplicationMaster and the executors.
-    if (sparkConf.contains("spark.submit.pyArchives")) {
-      var pythonPath = sparkConf.get("spark.submit.pyArchives")
-      if (env.contains("PYTHONPATH")) {
-        pythonPath = Seq(env.get("PYTHONPATH"), pythonPath).mkString(File.pathSeparator)
+    // If pyFiles contains any .py files, we need to add LOCALIZED_PYTHON_DIR to the PYTHONPATH
+    // of the container processes too. Add all non-.py files directly to PYTHONPATH.
+    //
+    // NOTE: the code currently does not handle .py files defined with a "local:" scheme.
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: should we add a warning when this happens?

**Comment by andrewor14** [comment]
@vanzin This looks great. I think the approach makes a lot of sense. The inline suggestions I have are mainly documentation and code organization.

A higher level comment I have is that there is currently no high-level documentation on how "Python is handled in YARN cluster mode", say vs client mode. To understand how it works the reader right now has to go through the code and look at how container envs are set up and read the comments there. I think it would be very useful to add a large block comment somewhere in `SparkSubmit` or `Client` to explain this on a higher level. Some things to potentially include in this block comment:
- pyspark / py4j files are automatically zipped up and put on the `PYTHONPATH` of all executors
- .py files are put in a special dir, which is then put on the `PYTHONPATH`
- archive python files (.egg, .zip etc.) are put directly on the `PYTHONPATH`
- primary python file is only distributed in cluster mode and only to the AM (because the driver runs there)
- where these python files are (both primary and secondary)

Another thing I was wondering is: was there a particular reason behind the change to how Spark properties are distributed (command line java opts -> `--properties-file`)? I think it's OK to keep it as is, but it would be good to put that in a java doc somewhere. (It's still a surprise to me how `Client` is a thousand line file but doesn't even have a high level java doc on the class)

TL;DR I think this is a good change for master. It's pretty close to being merged aside from a few documentation suggestions.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 2 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Ok, I'll revert to the old style. I actually would like to move the YARN code to Client.scala but didn't want to pile that up onto this change.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 6 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: Because PythonRunner needs that argument to be present.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 7 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I don't really know. That was part of the previous code.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 9 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I don't think it will ever happen since this is the first call to `distribute`, so it's mostly paranoia.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 11 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Nested methods cannot have modifiers.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 19 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: From the change description:

> Because the Client initialization code is a maze of implicit dependencies, some code needed to be moved around to make sure all needed state was available when the code ran.

So yes, I have verified it.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 20 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: This is a general problem with pyspark in general (not handling local: URIs), not particular to YARN...

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 18 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: I'll reword it.

**Comment by vanzin** [comment]
> A higher level comment I have is that there is currently no high-level documentation on how "Python is handled in YARN cluster mode", say vs client mode.

I think the main goal is to not have a difference between YARN client vs. cluster mode. This change goes a long way towards that goal.

**Comment by vanzin** [comment]
> Another thing I was wondering is: was there a particular reason behind the change to how Spark properties are distributed (command line java opts -> --properties-file)?

Mainly because if you do it in the command line YARN will replace any env variable references in the values, which in this case we do not want (because we want the expansion to occur on the node where executors will be launched, no where the AM is run).

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 2 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: yeah that might be a good change

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 6 above)
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: can you add comment

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 11 above)
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: oh I didn't even realize this is in a method... this YARN code is monstrous

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 18 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Actually, re-tested without this and it seems like using the property file handles the original issue already. So I'll just remove it.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 7 above)
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: (when you come back from your vacation) @tgravescs do you know?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 2 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: I ended up doing that in this change, since I needed to modify SparkSubmit anyway because the two conditions needed to run in the opposite order... well, long story short, the YARN code is now in Client.scala, where it should have always been.

**Comment by vanzin** [comment]
Jekins, retest this please.

**Comment by vanzin** [comment]
Hmm, moving the archive-finding code to Client.scala caused the failure... looking.

**Comment by srowen** [comment]
@andrewor14 how's this looking to you? I think the review feedback was incorporated. Do we need @tgravescs to have a look?

**Comment by andrewor14** [comment]
I'll take a look again later today, but last time I checked this is already in a good state.

**Comment by tgravescs** [comment]
no need to wait for me. I had taken a quick look and think the overall approach looks good.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 21]
```diff
@@ -503,8 +513,8 @@ private[spark] class ApplicationMaster(
     val userThread = new Thread {
       override def run() {
         try {
-          val mainArgs = new Array[String](args.userArgs.size)
-          args.userArgs.copyToArray(mainArgs, 0, args.userArgs.size)
+          val mainArgs = new Array[String](userArgs.size)
+          userArgs.copyToArray(mainArgs, 0, userArgs.size)
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: can't this just be `val mainArgs = userArgs.toArray`? Am I missing something?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 22]
```diff
@@ -490,9 +498,11 @@ private[spark] class ApplicationMaster(
         new MutableURLClassLoader(urls, Utils.getContextOrSparkClassLoader)
       }
 
+    var userArgs = args.userArgs
     if (args.primaryPyFile != null && args.primaryPyFile.endsWith(".py")) {
-      System.setProperty("spark.submit.pyFiles",
-        PythonRunner.formatPaths(args.pyFiles).mkString(","))
+      // Second argument is the list of files to add to PYTHONPATH, which Client.scala already
+      // handles, so it's empty.
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: It took a while for me to get this. This is actually referring to the arguments to `o.a.s.deploy.PythonRunner`'s main method. We need to document this because others who are not familiar with Spark submit at all will most certainly get lost around here.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala (hunk 21 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: This comes from all the way back when there was `yarn/alpha` and `yarn/stable`, but yeah, it doesn't look necessary.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 23]
```diff
@@ -548,8 +606,19 @@ private[spark] class Client(
     logInfo("Setting up container launch context for our AM")
     val appId = newAppResponse.getApplicationId
     val appStagingDir = getAppStagingDir(appId)
-    val localResources = prepareLocalResources(appStagingDir)
-    val launchEnv = setupLaunchEnv(appStagingDir)
+    val pySparkArchives =
+      if (sys.props.getOrElse("spark.yarn.isPython", null) == "true") {
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: could this be `sys.props.getOrElse("...", "false").toBoolean` so it handles cases as well?

**Comment by andrewor14** [comment]
Ok, I tested this on a Hadoop 2.4 cluster and verified that both client and cluster modes work with `--py-files`. I'll merge this once you address the last round of comments.

**Comment by andrewor14** [comment]
One higher level thing I'd like to mention is that the YARN code is getting quite difficult to review. We cleaned up a lot of the duplicate code and added documentation everywhere at one point, and that helped a lot. However, it is still difficult to follow the code or ensure its correctness because of a few things:

- Monstrous 100-200 lines methods like `setupLaunchEnv` and `prepareLocalResources`. These will only grow over time.
- Lack of high level documentation of what important classes like `Client`, `ApplicationMaster` and `ExecutorRunnable` do. Most of these don't even have class-level java docs.
- There are many implicit assumptions (e.g. use URI fragments if the local scheme is present) that are carried over from legacy code and undocumented
- Test coverage is poor. This is related to the first point; having large methods makes it difficult to write tests at fine granularity.

Rather than suggesting that we fix all these at once, I'm just jotting down my thoughts so we don't forget to fix these down the road. I'll add these thoughts to the umbrella YARN clean up JIRA.

**Comment by vanzin** [comment]
> the YARN code is getting quite difficult to review

Yeah, that comes from a long time. It really needs some cleaning up. It's not just the size of methods, but the implicit dependencies they have on one another (method A sets some instance field that is needed by method B, so they must be executed in that order or things break). That's the trickiest thing when messing with this code (the size is not that big of an issue IMO).

**Comment by andrewor14** [comment]
OK, I've updated [SPARK-3492](https://issues.apache.org/jira/browse/SPARK-3492). Yes, all implicit dependencies should be documented or even removed.

**Comment by andrewor14** [comment]
LGTM merging into master thanks @vanzin.

**Comment by zjffdu** [comment]
@vanzin  I am reading the yarn related code specially on org.apache.spark.deploy.yarn.Client.scala
Do you know where LOCAL_SCHEME("local") come from ? As I know we use file:// to represent a local resource, so not sure where "local" come from. 

Another question is that if I specify spark.yarn.jar as a hdfs location, the yarn client will still copy it to staging directory, I don't know why we do this. Would just use the hdfs file as LocalResource without copying much easier ?

**Comment by vanzin** [comment]
@zjffdu please don't comment on closed PRs, especially ancient ones; that's super confusing. Use the dev@ list for that.

To answer your questions: the "local:" scheme is defined in Spark's documentation. Defining a "hdfs" location for `spark.yarn.jar` should not cause it to be copied to the staging dir, unless they are on different HDFS servers. Maybe the way you're defining the value is confusing the code in Client.scala that checks that.

**Comment by zjffdu** [comment]
Thanks @vanzin, my fault, I specify the hdfs location using hostname, while it is ip address in core-site.xml (anyway, maybe we can improve it here)
